### PR TITLE
This feature adds a tag attribute to a snapshot. 

### PR DESCRIPTION
**Pull request by vpramo**
Using this users can,mark the snapshot with a user-readable tag, which can then be used to identify the snapshot
For eg: a snapshot which has passed all tests can be marked as "stable"

### CODE REVIEW

File: aasemble/django/apps/api/v2/serializers.py [hunk 1]
```diff
@@ -49,11 +49,11 @@ class Meta:
 
 class SnapshotSerializer(serializers.HyperlinkedModelSerializer):
     self = serializers.HyperlinkedRelatedField(view_name='v2_snapshot-detail', read_only=True, source='*', lookup_field='uuid')
-    mirrorset = serializers.HyperlinkedRelatedField(view_name='v2_mirrorset-detail', read_only=True, lookup_field='uuid')
+    mirrorset = serializers.HyperlinkedRelatedField(view_name='v2_mirrorset-detail', queryset=mirrorsvc_models.MirrorSet.objects.all(), lookup_field='uuid')
```
Comment (sorenisanerd) [inline aasemble/django/apps/api/v2/serializers.py]: This is unrelated to this feature. Please remove it.

File: aasemble/django/apps/mirrorsvc/models.py [hunk 2]
```diff
@@ -125,6 +125,7 @@ class Snapshot(models.Model):
     uuid = models.UUIDField(unique=True, default=uuid.uuid4, editable=False)
     timestamp = models.DateTimeField(auto_now_add=True)
     mirrorset = models.ForeignKey(MirrorSet)
+    tag = models.CharField(max_length=200, default='default')
```
Comment (sorenisanerd) [inline aasemble/django/apps/mirrorsvc/models.py]: Why have a default tag?

**Comment by sorenisanerd** [comment]
I thinks we should be able to specify more than a single tag per snapshot.

File: aasemble/django/apps/api/v1/serializers.py [hunk 3]
```diff
@@ -53,7 +53,7 @@ class SnapshotSerializer(serializers.HyperlinkedModelSerializer):
 
     class Meta:
         model = mirrorsvc_models.Snapshot
-        fields = ('self', 'timestamp', 'mirrorset')
+        fields = ('self', 'timestamp', 'mirrorset', 'tag')
```
Comment (sorenisanerd) [inline aasemble/django/apps/api/v1/serializers.py]: Oh, and I don't think we should add more features to the v1 API. Please only add this to the v2 API.

File: aasemble/django/apps/mirrorsvc/models.py (hunk 2 above)
Comment (vpramo) [inline aasemble/django/apps/mirrorsvc/models.py]: ok i get your point, since this field is not optional , people got to fill it with some tag initially

File: aasemble/django/apps/api/v1/serializers.py (hunk 3 above)
Comment (vpramo) [inline aasemble/django/apps/api/v1/serializers.py]: ok i will remove this from v1

**Comment by vpramo** [comment]
@sorenh done with the changes, please review, i will then squash the commits

File: aasemble/django/apps/mirrorsvc/models.py [hunk 4]
```diff
@@ -125,6 +125,7 @@ class Snapshot(models.Model):
     uuid = models.UUIDField(unique=True, default=uuid.uuid4, editable=False)
     timestamp = models.DateTimeField(auto_now_add=True)
     mirrorset = models.ForeignKey(MirrorSet)
+    tag = models.CharField(max_length=200)
```
Comment (sorenisanerd) [inline aasemble/django/apps/mirrorsvc/models.py]: I think it's better if we make it a separate model with a ForeignKey back to Snapshot. That way, we can more efficiently search for Snapshots that have a given tag.

Comment (vpramo) [inline aasemble/django/apps/mirrorsvc/models.py]: @sorenh ok, but if i create a seperate model, with tag and snapshot as foreign key, dont i need to create another api-endpoint to fill in the tag

Comment (sorenisanerd) [inline aasemble/django/apps/mirrorsvc/models.py]: No, you can expose them however you please. The django-rest-framework docs have several examples.

File: aasemble/django/apps/api/v2/views.py [hunk 5]
```diff
@@ -68,6 +68,7 @@ def perform_create(self, serializer):
 class SnapshotViewSet(mixins.CreateModelMixin,
                       mixins.RetrieveModelMixin,
                       mixins.DestroyModelMixin,
+                      mixins.UpdateModelMixin,
```
Comment (sorenisanerd) [inline aasemble/django/apps/api/v2/views.py]: You can replace all of these with `class SnapshotViewSet(viewsets.ModelViewSet)`. I listed these individually to avoid UpdateModelMixin, since I didn't want to allow changing anything about snapshots.

**Comment by sorenisanerd** [comment]
Can you please update the API documentation to show how to use this feature?

**Comment by sorenisanerd** [comment]
Can you please add some API tests to verify that this works as expected? Also add a test to make sure neither timestamp nor mirrorset can be modified, now that UpdateMixin is back.

**Comment by sorenisanerd** [comment]
@vpramo Let me know if you need any help.

**Comment by vpramo** [comment]
@sorenh will be working today and submit the tests

**Comment by sorenisanerd** [comment]
Sweet, thanks!

**Comment by vpramo** [comment]
@sorenh please check the test cases, tell me if i am in the right direction

File: aasemble/django/apps/api/tests.py [hunk 6]
```diff
@@ -306,3 +305,64 @@ def test_github_push_event(self, poll_one):
...
+    def test_create_mirrorset(self):
+        data = {'url': 'http://example.com/',
+                'series': ['trusty'],
+                'components': ['main']}
+        authenticate(self.client, 'eric')
+        response = self.client.post(self.mirror_url, data, format='json')
+        self.assertEquals(response.status_code, 201)
+        mirror = response.data
+        data = {'mirrors': [mirror['self']]}
+        authenticate(self.client, 'eric')
```
Comment (sorenisanerd) [inline aasemble/django/apps/api/tests.py]: There are already snapshots in the database you can work with.

**Comment by vpramo** [comment]
@sorenh hopefully this should complete it :)

**Comment by mayankkapoor** [comment]
@vpramo Rebase is required now, and travis CI will need to pass before this can be merged. Almost there.

File: aasemble/django/apps/mirrorsvc/models.py [hunk 7]
```diff
@@ -163,3 +163,21 @@ def save(self, *args, **kwargs):
...
+
+    class meta:
+        unique_together = ('snapshot', 'tag',)
+
+    def save(self, *args, **kwargs):
+        self.tag = self.tag.strip()
+        super(Tags, self).save(*args, **kwargs)
+
+    def user_can_modify(self, user):
+        return user == self.snapshot.mirrorset.owner
```
Comment (sorenisanerd) [inline aasemble/django/apps/mirrorsvc/models.py]: I'd suggest instead invoking `self.snapshot.user_can_modify(user)`. That way, if we decide to change the logic for allowing edits, we'd only need to edit it in one place.

File: aasemble/django/apps/api/tests.py [hunk 8]
```diff
@@ -306,3 +305,43 @@ def test_github_push_event(self, poll_one):
...
+        response = self.client.patch(self.list_url + '470688a8-7294-4c17-b020-1d67aebaf972/', data, format='json')
+        self.assertEquals(response.status_code, 200)
+        data['self']= response.data['self']
+        data['timestamp'] = response.data['timestamp']
+        data['mirrorset'] = response.data['mirrorset']
+        self.assertEquals(data, response.data)
+        return response.data
+
+    def test_update_snapshot_mirrorset_403(self):
```
Comment (sorenh) [inline aasemble/django/apps/api/tests.py]: Why should this give a 403? I don't disagree, but from an outsider's perspective, it's not clear from the method name or a comment or anything else that what you're trying to do is change the snapshot's mirrorset attribute to something else than its original value (or so I'm guessing, I didn't actually check).

In other words: Make the test name more descriptive.

File: aasemble/django/apps/api/tests.py [hunk 9]
```diff
@@ -306,3 +305,43 @@ def test_github_push_event(self, poll_one):
         from .tasks import github_push_event
         github_push_event("https://github.com/eric/project0")
         poll_one.delay.assert_called_with(1)
+
+class APIv2TagsTests(APIv1Tests):
+    list_url = '/api/v2/snapshots/'
+
+    def test_create_snapshot_tag(self):
+        data = {'mirrorset': 'http://testserver/api/v2/mirror_sets/60d0ba66-d343-404b-a6e6-5c141db11a54/',
+                'tags':[{'tag':'non-default'}]}
```
Comment (sorenisanerd) [inline aasemble/django/apps/api/tests.py]: This feels awkward.

If there are multiple tags, would I provide something like: `'tags': [{'tag': 'firsttag'}, {'tag': 'secondtag'}]`?

It really should just be `'tags': ['firsttag', 'secondtag']`.

File: aasemble/django/apps/mirrorsvc/models.py (hunk 7 above)
Comment (vpramo) [inline aasemble/django/apps/mirrorsvc/models.py]: sure will do that

File: aasemble/django/apps/api/tests.py (hunk 9 above)
Comment (vpramo) [inline aasemble/django/apps/api/tests.py]: ok i was just testing few things , i will add the same as mentioned

**Comment by vpramo** [comment]
@sorenh i am not able to find the comment on Error 403, i can do the following , i can do two things 
1. send 200 with a reply saying  "Update of mirrorset/timestamp is disallowed"
2. Send 403 with the above (not sure if that is possible , but can try)

**Comment by sorenisanerd** [comment]
@vpramo, I think it should probably be an HTTP 400. That seems to be consistent with the rest of django-rest-framework. You can raise a rest_framework.exceptions.ValidationError.

File: aasemble/django/apps/mirrorsvc/models.py [hunk 10]
```diff
@@ -163,3 +163,21 @@ def save(self, *args, **kwargs):
...
+
+    class meta:
+        unique_together = ('snapshot', 'tag',)
+
+    def save(self, *args, **kwargs):
+        self.tag = self.tag.strip()
+        super(Tags, self).save(*args, **kwargs)
+
+    def user_can_modify(self, user):
+        return user == self.snapshot.user_can_modify(user)
```
Comment (sorenisanerd) [inline aasemble/django/apps/mirrorsvc/models.py]: Just make it `return self.snapshot.user_can_modify(user)`. Right now, you're comparing `user` to either `True` or `False`.

File: aasemble/django/apps/api/tests.py [hunk 11]
```diff
@@ -500,6 +499,45 @@ def test_github_push_event(self, poll_one):
         github_push_event("https://github.com/eric/project0")
         poll_one.delay.assert_called_with(1)
 
+class APIv2TagsTests(APIv1Tests):
+    list_url = '/api/v2/snapshots/'
+
+    def test_create_snapshot_tag(self):
+        data = {'mirrorset': 'http://testserver/api/v2/mirror_sets/60d0ba66-d343-404b-a6e6-5c141db11a54/',
+                'tags':[{'tag':'firsttag'}, {'tag':'secondtag'}]}
```
Comment (sorenisanerd) [inline aasemble/django/apps/api/tests.py]: You're still passing `{'tags': [{'tag': 'firsttag'}, {'tag': 'secondtag'}]}`

It needs to be: `{'tags': ['firsttag', 'secondtag']}`

**Comment by piyush110786** [comment]
@vpramo .. your flakes test are failing please have a look in travis-ci and again try to push the code... :-)

**Comment by sorenisanerd** [comment]
Great job! :)

**Comment by mayankkapoor** [comment]
@vpramo Awesome job
